src/domain/human-description/entities/human-description-template.entity.ts

```diff
@@ -23,11 +31,29 @@ export class HumanDescriptionTemplate {
   private static readonly REGEX =
     /{{(?<typeToken>\w+) \$(?<paramIndex>\d+)}}|(?<wordToken>\w+)/g;
 
+  /**
+   * Store the regex matches as an array instead of an iterable so that it can be restarted
+   * @private
+   */
+  private readonly templateMatches: SafeRegExpMatchArray[];
+
   constructor(
     functionSignature: string,
     private readonly template: string,
   ) {
     this.functionAbi = parseAbi([functionSignature]);
+
+    this.templateMatches = Array.from(
+      template.matchAll(HumanDescriptionTemplate.REGEX),
+    )
+      .map((match) => {
+        if (!match.groups) {
+          throw Error(`Error parsing template ${this.template}`);
+        }
+
+        return match;
+      })
+      .filter(hasGroups);
```

Comment:
The `map` above checks if the matches have groups but then we call `hasGroups` again here.

Shouldn't the filter tackle this for us beforehand?

Reply:
This is an issue with Typescript not being able to filter out undefined values so we have to use a type-guard. Alternatively, we could also cast the value but I think its less clean.